File: openai_compatible_examples/basic_inference/openai_sdk_json.py
import os
import json
import sys
from openai import OpenAI, APIError
from dotenv import load_dotenv

# Add the parent directory (openai_compatible_examples) to sys.path
# to allow importing from the 'utils' module
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.append(parent_dir)

from utils.auth_helpers import get_api_key

# Load environment variables from .env file
load_dotenv()

# Get API details from environment variables
api_base_url = os.getenv("OPENAI_API_BASE")
api_key = os.getenv("OPENAI_API_KEY", "dummy-key") # Default to a dummy key if not set
model_name = os.getenv("MODEL_NAME", "default-model") # Provide a default model name

if not api_base_url:
    raise ValueError("OPENAI_API_BASE environment variable not set.")

print(f"--- Configuring OpenAI SDK ---")
print(f"Base URL: {api_base_url}")
print(f"Model: {model_name}")
print("API Key: Using provided key (or dummy key)")
print("---")

# Define the messages payload, explicitly asking for JSON
messages = [
    {
        "role": "system",
        "content": "You are a helpful assistant designed to output JSON."
    },
    {
        "role": "user",
        "content": "Provide details for a fictional product: Name, Category, and Price. Respond ONLY with a valid JSON object."
    }
]

def main():
    print("--- Sending JSON mode request using OpenAI SDK ---")
    print(f"Base URL: {api_base_url}")
    print(f"Model: {model_name}")
    print("NOTE: Ensure your prompt instructs the model to output JSON.")
    print("      And that your server supports response_format.")
    print("---")

    try:
        # Initialize client - API key is fetched dynamically *per request* below
        client = OpenAI(
            base_url=api_base_url,
            api_key="temp-key" # Initial key, will be replaced
        )

        # Fetch the latest API key and update the client
        client.api_key = get_api_key()

        # Send the request
        chat_completion = client.chat.completions.create(
            model=model_name,
            messages=messages,
            response_format={ "type": "json_object" }, # Request JSON mode
            max_tokens=100,
            temperature=0.5,
        )

        print("--- Full API Response Object ---")
        print(chat_completion.model_dump_json(indent=2))
        print("---")

        # Extract the message content (should be a JSON string)
        if chat_completion.choices:
            message = chat_completion.choices[0].message
            if message and message.content:
                message_content_str = message.content
                print(f"Assistant's raw content (should be JSON string):\n{message_content_str}")
                print("---")
                # Attempt to parse the message content string as JSON
                try:
                    parsed_content_json = json.loads(message_content_str)
                    print("Parsed JSON content from Assistant:")
                    print(json.dumps(parsed_content_json, indent=2))
                except json.JSONDecodeError:
                    print("Could not parse the assistant's message content as JSON.")
            else:
                print("Could not find message content in the response.")
        else:
            print("No choices found in the response.")

    except APIError as e:
        # Handle API errors
        print(f"An API error occurred: {e}")
        print(f"Status Code: {e.status_code}")
        print(f"Response: {e.response}")
        raise # Re-raise the exception after printing details
    except Exception as e:
        # Handle other potential errors
        print(f"An unexpected error occurred: {e}")
        raise # Also re-raise other unexpected errors

if __name__ == "__main__":
    main() 

File: java_sdk_examples/src/main/java/com/example/basic_inference/OpenAiSdkJson.java
package com.example.basic_inference;

import com.example.utils.ConfigLoader;
import com.theokanning.openai.completion.chat.ChatCompletionRequest;
import com.theokanning.openai.completion.chat.ChatCompletionChoice;
import com.theokanning.openai.completion.chat.ChatMessage;
import com.theokanning.openai.completion.chat.ChatMessageRole;
import com.theokanning.openai.completion.chat.ChatCompletionRequest.ChatCompletionRequestFunctionCall;
import com.theokanning.openai.service.OpenAiService;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

// Imports for custom OkHttpClient and Retrofit (needed for custom base URL)
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.theokanning.openai.client.OpenAiApi;
import static com.theokanning.openai.service.OpenAiService.*;

public class OpenAiSdkJson {

    public static void main(String[] args) {
        String apiKey = ConfigLoader.getApiKey();
        String apiBaseUrl = ConfigLoader.getApiBaseUrl();
        String modelName = ConfigLoader.getModelName();

        if ("DUMMY_API_KEY".equals(apiKey)) {
            System.err.println("API Key not configured. Please check your .env file.");
            return;
        }

        System.out.println("--- TheoKanning/openai-java SDK JSON Mode Chat Completion --- ");

        OpenAiService service;
        try {
            if (apiBaseUrl != null && !apiBaseUrl.equals("https://api.openai.com/v1/") && !apiBaseUrl.isEmpty()) {
                System.out.println("Using custom API base URL: " + apiBaseUrl);
                ObjectMapper mapper = defaultObjectMapper();
                OkHttpClient client = defaultClient(apiKey, Duration.ofSeconds(30))
                    .newBuilder()
                    .build();
                Retrofit retrofit = defaultRetrofit(client, mapper)
                    .newBuilder()
                    .baseUrl(apiBaseUrl)
                    .build();
                OpenAiApi openAiApi = retrofit.create(OpenAiApi.class);
                service = new OpenAiService(openAiApi);
            } else {
                System.out.println("Using default OpenAI API base URL.");
                service = new OpenAiService(apiKey, Duration.ofSeconds(30));
            }
        } catch (Exception e) {
            System.err.println("Error initializing OpenAiService: " + e.getMessage());
            e.printStackTrace();
            return;
        }

        List<ChatMessage> messages = new ArrayList<>();
        messages.add(new ChatMessage(ChatMessageRole.SYSTEM.value(), "You are a helpful assistant designed to output JSON."));
        messages.add(new ChatMessage(ChatMessageRole.USER.value(), "Provide a JSON object with two keys: 'name' and 'city', for a fictional character."));

        ChatCompletionRequest.Builder requestBuilder = ChatCompletionRequest.builder()
                .messages(messages)
                .temperature(0.7)
                .n(1);
                // For JSON mode, the API typically expects a `response_format` parameter.
                // The TheoKanning SDK might not have a direct setter for `response_format` in ChatCompletionRequest.builder().
                // It might be set via a generic parameter map or might require a newer SDK version / different method.
                // For now, we proceed without explicitly setting it, relying on the system message and prompt structure.
                // If the specific endpoint *requires* `{"type": "json_object"}`, this might not work as expected
                // and would need a way to add arbitrary top-level parameters to the request, or SDK update.
                // requestBuilder.responseFormat(Map.of("type", "json_object")); // Hypothetical setter

        if (modelName != null && !modelName.isEmpty()) {
            requestBuilder.model(modelName);
        } else {
            System.out.println("No model specified, using default 'gpt-3.5-turbo'. This model should support JSON mode if prompted correctly.");
            requestBuilder.model("gpt-3.5-turbo"); // Or a model known to be good with JSON output
        }
        ChatCompletionRequest request = requestBuilder.build();

        System.out.println("Sending JSON mode request...");
        System.out.println("Model: " + request.getModel());
        System.out.println("Messages: " + request.getMessages().stream().map(m -> m.getRole() + ": " + m.getContent()).collect(Collectors.joining("\n")) );
        System.out.println("------------------------------------");

        try {
            List<ChatCompletionChoice> choices = service.createChatCompletion(request).getChoices();

            System.out.println("--- API Response (JSON Mode Attempt) --- ");
            if (choices != null && !choices.isEmpty()) {
                ChatMessage assistantMessage = choices.get(0).getMessage();
                String responseContent = assistantMessage.getContent();
                System.out.println("Assistant Message (Raw):\n" + responseContent);

                // Attempt to parse as JSON to verify
                try {
                    ObjectMapper jsonParser = new ObjectMapper();
                    Object jsonObj = jsonParser.readValue(responseContent, Object.class); // Parses into a generic Map/List structure
                    System.out.println("Successfully parsed as JSON:");
                    System.out.println(jsonParser.writerWithDefaultPrettyPrinter().writeValueAsString(jsonObj));
                } catch (JsonProcessingException e) {
                    System.err.println("Could not parse assistant message as JSON: " + e.getMessage());
                    System.err.println("The model might not have returned valid JSON despite the prompt.");
                }
            } else {
                 System.out.println("No choices returned in the response.");
            }
            System.out.println("------------------------------------");

        } catch (retrofit2.HttpException e) {
            System.err.println("API HTTP Error occurred: " + e.code() + " " + e.message());
            try {
                String errorBody = e.response().errorBody() != null ? e.response().errorBody().string() : "No error body";
                System.err.println("Error Body: " + errorBody);
            } catch (Exception ioException) {
                System.err.println("Could not parse error body: " + ioException.getMessage());
            }
            e.printStackTrace();
        } catch (Exception e) {
            System.err.println("An unexpected error occurred: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("SDK JSON mode example complete.");
        service.shutdownExecutor();
    }
} 